Add blog post describing PSP deprecation and next steps. (#27369)
kubernetes/website: content/en/blog/_posts/2021-04-08-PodSecurityPolicy-Past-Present-and-Future.md

---
layout: blog
title: "PodSecurityPolicy: Past, Present, and Future"
date: 2021-04-08
slug: PodSecurityPolicy-Past-Present-and-Future
---

**Author:** Tabitha Sable (Kubernetes SIG Security)

> PodSecurityPolicy (PSP) is being deprecated in Kubernetes 1.21. This starts the countdown to its removal, but doesn’t change anything else. PodSecurityPolicy will continue to be fully functional for several more releases before being removed completely. In the meantime, we are developing a replacement for PSP that covers key use cases more easily and sustainably.

Reviewer: PodSecurityPolicy is _also_ the `kind` of an object in the API you use to configure that admission controller's behavior. Some people will only have seen the API end of things, not the kube-apiserver configuration that makes it work.

What are Pod Security Policies? Why did we need them? Why are they going away, and what’s next? How does this affect you? These key questions come to mind as we prepare to say goodbye to PSP, so let’s walk through them together. We’ll start with an overview of how features get removed from Kubernetes.

## What does deprecation mean in Kubernetes?

Whenever a Kubernetes feature is set to go away, our [deprecation policy](/docs/reference/using-api/deprecation-policy/) is our guide. First the feature is marked as deprecated, then after enough time has passed, it can finally be removed.

> Kubernetes 1.21 starts the deprecation process for PodSecurityPolicy. As with all feature deprecations, PodSecurityPolicy will continue to be fully functional for several more releases. The current plan is to remove PSP from Kubernetes in the 1.25 release.

Reviewer: If this article goes out before the v1.21 release, let's say something along the lines of “due for release later this week”.

Until then, PSP is still PSP. There will be at least a year during which the newest Kubernetes releases will still support PSP, and nearly two years until PSP will pass fully out of all supported Kubernetes versions.

## What is PodSecurityPolicy?

[PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) is a built-in [admission controller](/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/) that allows a cluster administrator to control security-sensitive aspects of the Pod specification. 

First, one or more PodSecurityPolicy resources are created in a cluster to define the requirements Pods must meet. Then, RBAC rules are created to control which PodSecurityPolicy applies to a given pod. If a pod meets the requirements of its PSP, it will be admitted to the cluster as usual. In some cases, PSP can also modify Pod fields, effectively creating new defaults for those fields. If a Pod does not meet the PSP requirements, it is rejected, and cannot run. 

> One more important thing to know about PodSecurityPolicy: it’s not the same as [PodSecurityContext](/docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context).

Reviewer: ```suggestion
One more important thing to know about PodSecurityPolicy: it’s not the same as the [`securityContext`](/docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context) for a Pod or container.
```

Author: Given that the YAML is `securityContext` but the API name is `PodSecurityContext`, and it's the API name that is confusingly similar looking to `PodSecurityPolicy`, I think this edit obscures the meaning of the warning/note here.

A part of the Pod specification, PodSecurityContext is the collection of fields that specify many of the security-relevant settings for a Pod. The security context dictates to the kubelet and container runtime how the Pod should actually be run. In contrast, the PodSecurityPolicy only constrains (or defaults) the values that may be set on the security context.

The deprecation of PSP does not affect PodSecurityContext in any way.

## Why did we need PodSecurityPolicy?

In Kubernetes, we define resources such as Deployments, StatefulSets, and Services that represent the building blocks of software applications. The various controllers inside a Kubernetes cluster react to these resources, creating further Kubernetes resources or configuring some software or hardware to accomplish our goals.

> In most Kubernetes clusters, RBAC (Role-Based Access Control) rules control access to these resources. `list`, `get`, `create`, `edit`, and `delete` are the sorts of API operations that RBAC cares about, but _RBAC does not consider what settings are being put into the resources it controls_. For example, a Pod can be almost anything from a simple webserver to a privileged command prompt offering full access to the underlying server node and all the data. It’s all the same to RBAC: a Pod is a Pod is a Pod.

Reviewer: _optionally_: link to https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole

## Why is PodSecurityPolicy going away?

In the years since PodSecurityPolicy was first introduced, we have realized that PSP has some serious usability problems that can’t be addressed without making breaking changes.

The way PSPs are applied to Pods has proven confusing to nearly everyone that has attempted to use them. It is easy to accidentally grant broader permissions than intended, and difficult to inspect which PSP(s) apply in a given situation. The “changing Pod defaults” feature can be handy, but is only supported for certain Pod settings and it’s not obvious when they will or will not apply to your Pod. Without a “dry run” or audit mode, it’s impractical to retrofit PSP to existing clusters safely, and it’s impossible for PSP to ever be enabled by default.

For more information about these and other PSP difficulties, check out SIG Auth’s KubeCon NA 2019 Maintainer Track session video: {{< youtube "SFtHRmPuhEw?start=953" youtube-quote-sm >}}

> Today, you’re not limited only to deploying PSP or writing your own custom admission controller. Several external admission controllers are available that incorporate lessons learned from PSP to provide a better user experience. [OPA/Gatekeeper](https://github.com/open-policy-agent/gatekeeper/), [Kyverno](https://github.com/kyverno/kyverno/), and [K-Rail](https://github.com/cruise-automation/k-rail) are all well-known, and each has its fans.

Reviewer: aside: CNCF policy on listing products that include 3rd party options is to write them in alphabetical order.

As this is a _blog post_ I think it's important to have regard for that policy but if we want to express a preference, we (specifically @tabbysable as article author) can absolutely do that.

Author: I'd actually like not to express a preference, thanks for calling out that policy.

---
layout: blog

> title: "PodSecurityPolicy: Past, Present, and Future"

Reviewer 2: This is a wonderful read! Tim covered the review, and I had just one minor addition from years of SEO work. 

This article will show up more often in Google/Bing/DDG searches and help more people if it's title and slug align to a searcher's intent. My assumption there is people will search `PSP deprecation` or PodSecurityPolicy deprecation`, so it'd be great to get those terms in both places. If you want to keep a similar format: `the How, What, and When of PSP Deprecation` could be fun. Or something playful like `Don't Panic: Updates on PodSecurityPolicy Deprecation`. Either way, the keyword in the title (and slug) helps a lot with ranking.

FWIW as well, I found this article is in the top position on search and it's got some great additional context. Adding here for others to see: https://www.antitree.com/2020/11/pod-security-policies-are-being-deprecated-in-kubernetes/

Reviewer 3: Thank you for the reminder about that blog post! It has some misinformation in it, so I sent a note to Mark to correct it. 😊

> date: 2021-04-08

Reviewer: If you're not sure about the date but that's the only thing to fix, you can add `draft: true` to the front matter and send in another PR to remove that once the date is confirmed.

> slug: PodSecurityPolicy-Past-Present-and-Future

Reviewer: ```suggestion
slug: podsecuritypolicy-past-present-and-future
```

Author: holding off on addressing this pending potential retitling

Reviewer: If changing the title, we might need to revise the Kubernetes v1.21 release blog to match.

Author: I've updated the title, please update the release blog to match.

Reviewer 4: Ack, done :)

---

**Author:** Tabitha Sable (Kubernetes SIG Security)

> PodSecurityPolicy (PSP) is being deprecated in Kubernetes 1.21. This starts the countdown to its removal, but doesn’t change anything else. PodSecurityPolicy will continue to be fully functional for several more releases before being removed completely. In the meantime, we are developing a replacement for PSP that covers key use cases more easily and sustainably.

Reviewer: Should we mention “released today” in connection with the Kubernetes 1.21 reference? The current article date is for the release day.

What are Pod Security Policies? Why did we need them? Why are they going away, and what’s next? How does this affect you? These key questions come to mind as we prepare to say goodbye to PSP, so let’s walk through them together. We’ll start with an overview of how features get removed from Kubernetes.

## What does deprecation mean in Kubernetes?

> Whenever a Kubernetes feature is set to go away, our [deprecation policy](docs/reference/using-api/deprecation-policy/) is our guide. First the feature is marked as deprecated, then after enough time has passed, it can finally be removed.

Reviewer: :warning: broken hyperlink
```suggestion
Whenever a Kubernetes feature is set to go away, our [deprecation policy](/docs/reference/using-api/deprecation-policy/) is our guide. First the feature is marked as deprecated, then after enough time has passed, it can finally be removed.
```

Kubernetes 1.21 starts the deprecation process for PodSecurityPolicy. As with all feature deprecations, PodSecurityPolicy will continue to be fully functional for several more releases. The current plan is to remove PSP from Kubernetes in the 1.25 release.

Until then, PSP is still PSP. There will be at least a year during which the newest Kubernetes releases will still support PSP, and nearly two years until PSP will pass fully out of all supported Kubernetes versions.

## What is PodSecurityPolicy?

> [PodSecurityPolicy](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) is a built-in [admission controller](https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/) that allows a cluster administrator to control security-sensitive aspects of the Pod specification. 

Reviewer: nit
```suggestion
[PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/) is a built-in [admission controller](o/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/) that allows a cluster administrator to control security-sensitive aspects of the Pod specification. 
```
We use site-relative hyperlinks.

First, one or more PodSecurityPolicy resources are created in a cluster to define the requirements Pods must meet. Then, RBAC rules are created to control which PodSecurityPolicy applies to a given pod. If a pod meets the requirements of its PSP, it will be admitted to the cluster as usual. In some cases, PSP can also modify Pod fields, effectively creating new defaults for those fields. If a Pod does not meet the PSP requirements, it is rejected, and cannot run. 

> One more important thing to know about PodSecurityPolicy: it’s not the same as PodSecurityContext.

Reviewer: ```suggestion
One more important thing to know about PodSecurityPolicy: it’s not the same as [PodSecurityContext](/docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context).
```
?

A part of the Pod specification, PodSecurityContext is the collection of fields that specify many of the security-relevant settings for a Pod. The security context dictates to the kubelet and container runtime how the Pod should actually be run. In contrast, the PodSecurityPolicy only constrains (or defaults) the values that may be set on the security context.

The deprecation of PSP does not affect PodSecurityContext in any way.

## Why did we need PodSecurityPolicy?

In Kubernetes, we define resources such as Deployments, StatefulSets, and Services that represent the building blocks of software applications. The various controllers inside a Kubernetes cluster react to these resources, creating further Kubernetes resources or configuring some software or hardware to accomplish our goals.

> In most Kubernetes clusters, RBAC (Role-Based Access Control) rules control access to these resources. Get, Create, Edit, and Delete are the sorts of operations that RBAC cares about, but RBAC does not consider what settings are being put into the resources it controls. For example, a Pod can be almost anything from a simple webserver to a privileged command prompt offering full access to the underlying server node and all the data. It’s all the same to RBAC: a Pod is a Pod is a Pod.

Reviewer: I'd write:
```suggestion
In most Kubernetes clusters, RBAC (Role-Based Access Control) rules control access to these resources. `list`, `get`, `create`, `edit`, and `delete` are the sorts of API operations that RBAC cares about, but _RBAC does not consider what settings are being put into the resources it controls_. For example, a Pod can be almost anything from a simple webserver to a privileged command prompt offering full access to the underlying server node and all the data. It’s all the same to RBAC: a Pod is a Pod is a Pod.
```
I added `list` to give a strong hint that these aren't HTTP verb names.

For more information about these and other PSP difficulties, check out SIG Auth’s KubeCon NA 2019 Maintainer Track session video: {{< youtube "SFtHRmPuhEw?start=953" youtube-quote-sm >}}

Today, you’re not limited only to deploying PSP or writing your own custom admission controller. Several external admission controllers are available that incorporate lessons learned from PSP to provide a better user experience. [OPA/Gatekeeper](https://github.com/open-policy-agent/gatekeeper/), [Kyverno](https://github.com/kyverno/kyverno/), and [K-Rail](https://github.com/cruise-automation/k-rail) are all well-known, and each has its fans.

Although there are other good options available now, we believe there is still value in having a built-in admission controller available as a choice for users. With this in mind, we turn toward building what’s next, inspired by the lessons learned from PSP.

## What’s next?

> Kubernetes SIG Security, SIG Auth, and a diverse collection of other community members have been working together for months to ensure that what’s coming next is going to be awesome. We have developed a Kubernetes Enhancement Proposal ([KEP 2579](https://github.com/kubernetes/enhancements/issues/2579)) and a prototype for a new feature, currently being called PSP Replacement Policy. We are targeting an Alpha release in Kubernetes 1.22.

Reviewer: Does it go without saying that _PSP Replacement Policy_ is a working title?

Author: I think that's pretty clearly implied by the "currently" but could also see changing to "new feature with a temporary working title of PSP Replacement Policy". Thoughts?

Reviewer: (whichever you prefer!)

PSP Replacement Policy is designed to be as simple as practically possible while providing enough flexibility to really be useful in production at scale. It has soft rollout features to enable retrofitting it to existing clusters, and is configurable enough that it can eventually be active by default. It can be deactivated partially or entirely, to coexist with external admission controllers for advanced use cases.

## What does this mean for you?

What this all means for you depends on your current PSP situation. If you’re already using PSP, there’s plenty of time to plan your next move. Please review the PSP Replacement Policy KEP and think about how well it will suit your use case.

If you’re making extensive use of the flexibility of PSP with numerous PSPs and complex binding rules, you will likely find the simplicity of PSP Replacement Policy too limiting. Use the next year to evaluate the other admission controller choices in the ecosystem. There are resources available to ease this transition, such as the [Gatekeeper Policy Library](https://github.com/open-policy-agent/gatekeeper-library).

> If your use of PSP is relatively simple, with a few policies and straightforward binding to service accounts in each namespace, you will likely find PSP Replacement Policy to be a good match for your needs. Evaluate your PSPs compared to the [Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards/) to get a feel for where you’ll be able to use the Restricted, Baseline, and Privileged policies. Please follow along with or contribute to the KEP and subsequent development, and try out the Alpha release of PSP Replacement Policy when it becomes available.

Reviewer: I'd write:
```suggestion
If your use of PSP is relatively simple, with a few policies and straightforward binding to service accounts in each namespace, you will likely find PSP Replacement Policy to be a good match for your needs. Evaluate your PSPs compared to the Kubernetes [Pod Security Standards](/docs/concepts/security/pod-security-standards/) to get a feel for where you’ll be able to use the Restricted, Baseline, and Privileged policies. Please follow along with or contribute to the KEP and subsequent development, and try out the Alpha release of PSP Replacement Policy when it becomes available.
```

> If you’re just beginning your PSP journey, you will save time and effort by keeping it simple. You can approximate the functionality of PSP Replacement Policy today by using the Pod Security Standards’ PSPs. If you set the cluster default by binding a Baseline or Restricted policy to the system:authenticated group, and then make a more-permissive policy available as needed in certain Namespaces [using ServiceAccount bindings](https://kubernetes.io/docs/concepts/policy/pod-security-policy/#run-another-pod), you will avoid many of the PSP pitfalls and have an easy migration to PSP Replacement Policy. If your needs are much more complex than this, your effort is probably better spent adopting one of the more fully-featured external admission controllers mentioned above.

Reviewer: ```suggestion
If you’re just beginning your PSP journey, you will save time and effort by keeping it simple. You can approximate the functionality of PSP Replacement Policy today by using the Pod Security Standards’ PSPs. If you set the cluster default by binding a Baseline or Restricted policy to the system:authenticated group, and then make a more-permissive policy available as needed in certain Namespaces [using ServiceAccount bindings](/docs/concepts/policy/pod-security-policy/#run-another-pod), you will avoid many of the PSP pitfalls and have an easy migration to PSP Replacement Policy. If your needs are much more complex than this, your effort is probably better spent adopting one of the more fully-featured external admission controllers mentioned above.
```
